modin/core/dataframe/pandas/partitioning/axis_partition.py

```diff
@@ -44,6 +95,99 @@ def list_of_blocks(self):
             partition.list_of_blocks[0] for partition in self.list_of_block_partitions
         ]
 
+    @property
+    def list_of_block_partitions(self) -> list:
+        """
+        Get the list of block partitions that compose this partition.
+
+        Returns
+        -------
+        List
+            A list of ``PandasDataframeAxisPartition``.
+        """
+        if self._list_of_block_partitions is not None:
+            return self._list_of_block_partitions
+        self._list_of_block_partitions = []
+        # Extract block partitions from the block and virtual partitions that
+        # constitute this partition.
+        for partition in self._list_of_constituent_partitions:
+            if isinstance(partition, PandasDataframeAxisPartition):
+                if partition.axis == self.axis:
+                    # We are building a virtual partition out of another
+                    # virtual partition `partition` that contains its own list
+                    # of block partitions, partition.list_of_block_partitions.
+                    # `partition` may have its own call queue, which has to be
+                    # applied to the entire `partition` before we execute any
+                    # further operations on its block parittions.
+                    partition.drain_call_queue()
+                    self._list_of_block_partitions.extend(
+                        partition.list_of_block_partitions
+                    )
+                else:
+                    # If this virtual partition is made of virtual partitions
+                    # for the other axes, squeeze such partitions into a single
+                    # block so that this partition only holds a one-dimensional
+                    # list of blocks. We could change this implementation to
+                    # hold a 2-d list of blocks, but that would complicate the
+                    # code quite a bit.
+                    self._list_of_block_partitions.append(
+                        partition.force_materialization().list_of_block_partitions[0]
+                    )
+            else:
+                self._list_of_block_partitions.append(partition)
+        return self._list_of_block_partitions
+
+    @classmethod
+    def _get_drain_func(cls):  # noqa: GL08
+        return PandasDataframeAxisPartition.drain
+
+    def drain_call_queue(self, num_splits=None):
+        """
+        Execute all operations stored in this partition's call queue.
+
+        Parameters
+        ----------
+        num_splits : int, default: None
+            The number of times to split the result object.
+        """
+        if len(self.call_queue) == 0:
+            # this implicitly calls `drain_call_queue` for block partitions,
+            # which might have deferred call queues
+            _ = self.list_of_blocks
+            return
+        call_queue = self.call_queue
+        try:
```

Comment:
What led you to change the logic here?

Reply:
I've changed this logic to preserve the old behavior, where we clear the call queue only after successfully draining it:
https://github.com/modin-project/modin/blob/46cb44bd0cc7cb599bc5cf4edea3cd13f10a35a2/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py#L508-L512

After I moved this condition to the base AxisPartition class, I was also forced to clear the call queue <i>before</i> actually calling the draining `.apply()` in order to avoid infinite recursion:
https://github.com/modin-project/modin/blob/46cb44bd0cc7cb599bc5cf4edea3cd13f10a35a2/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py#L374-L375
This made it possible to lose the call queue if an exception happens during the draining (this couldn't happen before, as we cleared the queue <i>after</i> successful draining), so I've just added this try-except safeguard in order to preserve the old behavior.